Fix memory corruption and leaks on clisp

vibs29 opened the pull request:
----
Files
  ffi-buffer.lisp
  ffi-buffer-clisp.lisp
  streams.lisp
  random.lisp
  x509.lisp

----
Memory corruption bug on clisp

s/b-replace
b/s-replace
had bugs that could cause them to miscalculate the buffer's end as being
beyond its boundary, or miscalculate the number of bytes to copy if the
buffer's end was specified but the sequence was smaller.
All callers of s/b-replace happened to pass arguments that didn't trigger
its bugs.
But one caller of b/s-replace (namely stream-write-sequence) could
legitimately call it in a way that did trigger one of its bugs.
E.g. if the buffer was smaller than the sequence, it would corrupt memory
by writing beyond the buffer's bounds.

I have fixed all the bugs, which were in s/b-replace and b/s-replace.

----
Performance

For clisp:

b/s-replace also copies less.
The old version always called subseq, which copies.
The new version copies only if the source seq is not a vector.

s/b-replace is not expected to allocate memory proportional to its
input arrays.
But due to its call to memory-as, it did.
Now it doesn't. It allocates O(1) memory, regardless of input array sizes.

b/s-replace also allocates O(1) memory now.

For all lisps:

stream-read-sequence
I have made this clearer.

stream-write-sequence
I have rewritten this to be clearer.
And faster. The old version could flush a non-full stream.
As a pathological case,
  writing 1 byte, then 2048 times writing 2049 bytes, would cause 4096
  flushes.
Now that will cause only 2049 flushes.

----
Memory leak on clisp

There was also a memory leak because foreign buffers were allocated but
never freed. I have fixed this by extending the buffer API and having
all callers of make-buffer also release the buffer when finished with it.

vibs29 committed:
Fix memory corruption and leaks on clisp

----
Files
  ffi-buffer.lisp
  ffi-buffer-clisp.lisp
  streams.lisp
  random.lisp
  x509.lisp

----
Memory corruption bug on clisp

s/b-replace
b/s-replace
had bugs that could cause them to miscalculate the buffer's end as being
beyond its boundary, or miscalculate the number of bytes to copy if the
buffer's end was specified but the sequence was smaller.
All callers of s/b-replace happened to pass arguments that didn't trigger
its bugs.
But one caller of b/s-replace (namely stream-write-sequence) could
legitimately call it in a way that did trigger one of its bugs.
E.g. if the buffer was smaller than the sequence, it would corrupt memory
by writing beyond the buffer's bounds.

I have fixed all the bugs, which were in s/b-replace and b/s-replace.

----
Performance

For clisp:

b/s-replace also copies less.
The old version always called subseq, which copies.
The new version copies only if the source seq is not a vector.

s/b-replace is not expected to allocate memory proportional to its
input arrays.
But due to its call to memory-as, it did.
Now it doesn't. It allocates O(1) memory, regardless of input array sizes.

b/s-replace also allocates O(1) memory now.

For all lisps:

stream-read-sequence
I have made this clearer.

stream-write-sequence
I have rewritten this to be clearer.
And faster. The old version could flush a non-full stream.
As a pathological case,
  writing 1 byte, then 2048 times writing 2049 bytes, would cause 4096
  flushes.
Now that will cause only 2049 flushes.

----
Memory leak on clisp

There was also a memory leak because foreign buffers were allocated but
never freed. I have fixed this by extending the buffer API and having
all callers of make-buffer also release the buffer when finished with it.

avodonosov commented:
re #120

avodonosov commented on src/x509.lisp:
What motivates the changes in this file, and what is the changes nature? Note, `decode-certificate` and `decode-certificate-from-file` are public functions.

vibs29 committed:
Revert x509.lisp because changes were incompatible

avodonosov commented:
How to avoid unnecessary changes in unmodified files is a secondary question. We will solve that after we agreed on the final version. 

But first I need to understand all the changes. I haven't digested your branch yet.

The comments in src/ffi-buffer-clisp.lisp say you got a significant speed up and attribute that to "copying via a single foreign call to MEMORY-AS instead of one foreign call per element via %MEM-REF.' Could you point where this mem-ref per element happens in the old code?

What whould be good to have is self-contained test cases demonstrating the bug in the old code and pass with the new code, and also covering all branches of the copying code changed and introduced. Do you see a sufficienty easy and practical way to implement such tests?

Are you open go have a call to help me understand the pull request?

vibs29 commented:
I'm unfamiliar with the bio code. I've taken a quick look at the write-puts test. I don't think what I've done will directly fix that. However, remembering that stream-write-sequence would on clisp illegally overwrite arbitrary parts of the C memory, if any of the tests called stream-write-sequence then all subsequent behaviour of that process is unpredictable. It'll certainly be worth trying the test suite of 163 to see if its problem has disappeared with this patch. And it's not worth reasoning about any misbehaviour on clisp prior to applying this patch. cl+ssl on clisp is dangerous and should never be used without this patch.

vibs29 committed:
Stabilise performance for byte-sized operations

stream-listen
stream-read-byte
are written to potentially have horrible performance.

The reason is that cffi's experimental Shareable Byte Vector interface
is bad. It indicates that although with-pointer-to-vector-data (wptvd)
with an empty body may be constant time (due to the vector's being shared
between Lisp and C), it may also be O(len(vector)) on implementations
that require copying in/out. Given that cffi is a portability layer,
the name make-shareable-byte-vector is misleading if the result isn't
necessarily shared between C and Lisp but may require copying. Anyway,
bad cffi names aside, callers must therefore treat wptvd as having
O(len(vector)) overhead since that is what its documentation indicates.

In that light, stream-listen and stream-read-byte, which are written to
potentially cause the copying of the entire underlying buffer when they
only want to operate on one byte, are badly written. They are written
with a mistaken assumption, that wptvd has O(1) overhead, but it is
really specified by cffi to have O(len(vector)) overhead.

One solution is for the stream to own another one byte input buffer for
these two functions to use. The input buffer is only used as a temporary
variable within a function. It doesn't retain meaningful state between
function calls. So this solution is easy to implement.

As it happens, with the clisp-specific ffi-buffer-clisp.lisp,
the clisp wptvd in cl+ssl is O(1) so there happens not to be a
problem presently. All current wptvd's in use by cl+ssl are O(1). But
stream-listen and stream-read-byte should still not be written to assume
that the general wptvd has O(1) overhead. A new wptvd may be added
to cffi or an old one rewritten to have the allowable O(len(vector))
overhead. Thus stream-listen and stream-read-byte should really be
corrected to reflect that wptvd does not guarantee that implementations
will have less than O(len(vector)) overhead. As written, they are just
lucky they happen to perform well because of the implementation details
of the current wptvd's in existence.

This commit implements the solution of having a separate small buffer
for stream-listen and stream-read-byte to use.

Performance will not improve today. But this design will ensure that
cl+ssl doesn't in the future suddenly mysteriously develop horrible
performance due to relying on implementation details of wptvd adding O(1)
overhead when it is really specified to add O(len(vector)) overhead.

vibs29 commented:
This is how I validate the last commit 45c439f "Stabilise performance for byte-sized operations". `*url*` needs editing first. Expect output of about one dot, lots of dashes, then a string form of the http response.
```
; insert the following traces into cl+ssl's streams.lisp to ensure it is called.
; stream-listen
;   (princ #\.) (finish-output)
; stream-read-byte
;   (princ #\-) (finish-output)

(defparameter *url* "https://...")

(defun test-byte ()
  "Exercises stream-listen and stream-read-byte"
  (multiple-value-bind (stream code headers uri stream2 must-close reason)
                       (drakma:http-request *url* :want-stream t)
    (unwind-protect
      (let ((str (chunga:chunked-stream-stream
                   (flexi-streams:flexi-stream-stream stream))))
        (print (list 'str-type (type-of str)))
        (do () ((listen str)))
        (princ
          (with-output-to-string (o)
            (handler-case
                (do ((b (read-byte str nil :eof)
                        (read-byte str nil :eof)))
                    ((eq b :eof))
                  (write-char (code-char b) o)) ;assume character
              (cl+ssl::ssl-error-ssl (e)
                (print e)
                ;SSL_get_error specifies not to call SSL_shutdown
                ;(but now your test leaks C buffer memory.)
                ;really, the error should be trapped by stream-read-byte
                (setf must-close nil))))))
      (when must-close
        (close stream))))
  (values))
```

avodonosov commented on src/streams.lisp:
```diff
@@ -112,7 +116,7 @@
         (setf (ssl-stream-peeked-byte stream) nil))
       (handler-case
-          (let ((buf (ssl-stream-input-buffer stream))
+          (let ((buf (ssl-stream-input-buffer-small stream))
```
Why reading into a single-byte buffer is better than reading one byte into the big buffer in stream-listen and stream-read-byte?

avodonosov replied:
Sorry I missed the commit message (I looked in the pull request conversation and in the code comments before asking though).

I see now in the commit message you adress that question.

vibs29 committed:
Don't inline non-trivial *-REPLACE functions

The original versions of s/b-replace and b/s-replace were one liners so
declared to be inlined, but not all versions are one-liners now so not
all versions should be inlined.

vibs29 committed:
Inline empty RELEASE-BUFFER

avodonosov commented on src/ffi-buffer-clisp.lisp:
```diff
@@ -58,40 +89,55 @@
-  (+ buf-start
-     (- (or seq-end (length seq))
-        seq-start)))
+(defparameter *mem-max* 1024 "so *-REPLACE require the expected O(1) memory")
```
The `*mem-max*` is intended to limit the allocations done by `memory-as`, right?  Question, if we copy 2048 bytes, calling memory-as twice for 1024 bytes, is it really better than a single call for 2048 bytes? The total amount of allocated memory is the same. If it is garbage collected, collecting two objects may be more work for GC than one object. Is it really beneficial?

Also, in context of cl+ssl, the maximum size of arrays copied with `s/b-replace`, `b/s-replace` is limited by the buffer size, which defaults to 2048 bytes. Not a big difference from the 1024 `*mem-max*` here.

Are you thinking of the `*-replace` functions as more general purpose utilities, that must be prepared to any size of collections copied?

avodonosov commented:
My comments about the Eventual Design Ideal:

The `cffi:with-pointer-to-vector-data` is not a bad API, it's just intended for a different use-case. Namely, for applications which want to call C functions like `read` or `SSL_read` directly, and have these functions to directly populate a lisp sequence, with zero intermediate copies. On lisp implementations that provide zero copying `cffi:with-pointer-to-vector-data` it is an ideal API for that case: give C function a native foreign pointer to the memory underlying a lisp sequence and then work with the sequence in Lisp using standard sequence functions.

In cl+ssl we have a different arrangement: we read into a buffer and then copy from the buffer to another sequence. So we may want better support in CFFI for this use case without deprecating `cffi:with-pointer-to-vector-data`.

Next, I think there is not need in CFFI to introduce an opaque buffer abstraction.

Applications can just use `foreign-alloc` and `foreign-free` for buffer allocation / release, `men-ref` and `mem-aref` for element-wise access.

The only thing missing is bulk copy between a foreign memory array and lisp sequence  (`b/s-relpace`, `s/b-replace` in terms of cl+ssl). If CFFI introduced functions for that, it will be feature complete for the use case we have in cl+ssl.

A big design question is how generic these bulk copy functions should be with respect to lisp sequence type: support only simple vectors, any vectors, or both vectors and lists. This depends on the possibilities we can reasonably expect from lisp implementations to efficiently provide.

avodonosov commented:
I undersdod what may be the motivation for opaque buffer abstraction in cffi: if efficient bulk co0ying operations are not available in all lisp, then some lisps may implement buffer using shareable-vector/wptvd and other lisps with foreign-alloc / bulk copy.

vibs29 committed:
Change internal defparameter to defconstant.

vibs29 committed:
Replace force-output with stream-force-output

vibs29 committed:
clisp: improve performance

s/b-replace had O(n^2) running time for an input seq of length n. Now
it is O(n).

Also, s/b-replace and b/s-replace now correctly raise errors when
bounding indices are out of bounds, whereas previously they would
sometimes effectively shrink illegal arguments to become legal, which
was unintentionally different from how replace behaves. Their behaviour
is modelled on replace's.

b/s-replace also wouldn't work for zero-length buffer/sequence. Now it
does. This bug caused no harm because it is never called with zero-length
buffer/sequence.

vibs29 committed:
clisp: minor refactoring
